File: src/console/command/parameter/def/PluginParameter.php
<?php

namespace pocketcloud\cloud\console\command\parameter\def;

use pocketcloud\cloud\console\command\parameter\CommandParameter;
use pocketcloud\cloud\console\command\parameter\exception\ArgumentParseException;
use pocketcloud\cloud\plugin\CloudPlugin;
use pocketcloud\cloud\plugin\CloudPluginManager;

readonly class PluginParameter extends CommandParameter {

    public function parseValue(string $input): CloudPlugin {
        if (($plugin = CloudPluginManager::getInstance()->get($input)) !== null) return $plugin;
        throw new ArgumentParseException();
    }

    public function onTabCompleteMatch(string $currentArg): array {
        return array_filter(array_keys(CloudPluginManager::getInstance()->getAll()), fn(string $plugin) => str_contains(strtolower($plugin), $currentArg));
    }

    public function getType(): string {
        return "server";
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/parameter/def/PluginParameter.java
package de.pocketcloud.cloud.console.command.parameter.def;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.command.exception.ArgumentParseException;
import de.pocketcloud.cloud.console.command.parameter.BaseCommandParameter;

import java.util.List;

public final class PluginParameter extends BaseCommandParameter {

    public PluginParameter(String name, boolean optional) {
        super(name, optional);
    }

    public PluginParameter(String name, boolean optional, String customErrorMessage) {
        super(name, optional, customErrorMessage);
    }

    @Override
    public Object parseValue(String input) throws ArgumentParseException {
        return PocketCloud.instance().plugins().get(input).orElseThrow(() -> new ArgumentParseException("Plugin not found"));
    }

    @Override
    public List<String> onTabCompleteMatch(String currentArg) {
        return PocketCloud.instance().plugins().getPlugins().values().stream().map(p -> p.getDescription().name()).toList();
    }

    @Override
    public String getType() {
        return "plugin";
    }
}